ocaml/xapi/repository_helpers.ml

```diff
@@ -717,6 +717,196 @@ let get_updates_from_repoquery repositories =
   in
   new_updates @ updates
 
+(* Parser module for the output of command yum upgrade *)
+module YumUpgradeOutput = struct
+  open Angstrom
+
+  let consume = Consume.Prefix
+
+  let is_eol = function '\n' -> true | _ -> false
+
+  let is_white = Astring.Char.Ascii.is_white
+
+  let rec line section section_acc acc =
+    let rec line_after_txn_summary sections =
+      at_end_of_input >>= function
+      | false -> (
+          take_till is_eol <* end_of_line >>= function
+          | l
+            when String.starts_with
+                   ~prefix:" yum load-transaction /tmp/yum_save_tx." l
+                 && String.ends_with ~suffix:".yumtx" l ->
+              return (sections, Some l)
+          | _ ->
+              line_after_txn_summary sections
+        )
+      | true ->
+          return (sections, None)
+    in
+    (* The following sections/separators come from
+     * https://github.com/rpm-software-management/yum/blob/master/output.py#L1569-L1576
+     *)
+    at_end_of_input >>= function
+    | false -> (
+        take_till is_eol <* end_of_line >>= function
+        | ( "Installing:"
+          | "Updating:"
+          | "Upgrading:"
+          | "Removing:"
+          | "Reinstalling:"
+          | "Downgrading:"
+          | "Installing for dependencies:"
+          | "Updating for dependencies:"
+          | "Removing for dependencies:" ) as section' -> (
+          match section with
+          | Some s ->
+              line (Some section') [] ((s, section_acc) :: acc)
+          | None ->
+              line (Some section') [] acc
+        )
+        | "Transaction Summary" -> (
+          match section with
+          | Some s ->
+              line_after_txn_summary ((s, section_acc) :: acc)
+          | None ->
+              line_after_txn_summary acc
+        )
+        | l -> (
```

Comment:
It seems it's collected in a different way (`Some s` instead of `Some l`, `l :: section_acc` instead of `[]`). This seems to have quite complex logic and some labels in the `line` function might help here

Reply:
Added the labels of parameters and some comments.
The structure of the result would be:
`[(section_name:string, lines: string list); (section_name:string, lines: string list); ...]`